pkg/controllers/machine/termination/controller.go

```diff
@@ -0,0 +1,125 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package termination
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"golang.org/x/time/rate"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/client-go/util/workqueue"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+var _ corecontroller.FinalizingTypedController[*v1alpha5.Machine] = (*Controller)(nil)
+
+// Controller is a Machine Termination controller that triggers deletion of the Node and the
+// CloudProvider Machine through its graceful termination mechanism
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor for the Machine Controller
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) corecontroller.Controller {
+	return corecontroller.Typed[*v1alpha5.Machine](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	})
+}
+
+func (*Controller) Name() string {
+	return "machine_termination"
+}
+
+func (c *Controller) Reconcile(_ context.Context, _ *v1alpha5.Machine) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	stored := machine.DeepCopy()
+	if !controllerutil.ContainsFinalizer(machine, v1alpha5.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	nodes, err := machineutil.AllNodesForMachine(ctx, c.kubeClient, machine)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if len(nodes) > 0 {
+		for _, node := range nodes {
+			if err := c.kubeClient.Delete(ctx, node); client.IgnoreNotFound(err) != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{}, nil
+	}
```

Comment:
This if statement is redundant, right? Since the for loop will check this anyway?

Actually, I guess if you want to not move on if there are still underlying nodes, this makes sense. Consider just checking it after the for loop anyway. 
```suggestion
	for _, node := range nodes {
		if err := c.kubeClient.Delete(ctx, node); client.IgnoreNotFound(err) != nil {
			return reconcile.Result{}, err
		}
	}
	if len(nodes) > 0 {
	        return reconcile.Result{}, nil
	}
```

Reply:
Additionally I think it'd be nice to add a comment here to indicate that you're relying on the node termination controller here. Something  like

```
// If there are underlying nodes, let the node termination controller handle graceful termination before continuing. 
```

Reply:
Fair point, I can see this being confusing